crates/bevy_animation/src/blend.rs

```diff
@@ -0,0 +1,189 @@
+pub trait Blend: Interpolate {
+    /// The default value of the blend, which has no effect when blended with other values.
+    const IDENTITY: Self;
+
+    /// Addatively blends another value on top of this one.
+    fn blend(self, other: Self, blend_weight: f32) -> Self;
+}
+
+/// The state for an ongoing blend operation. On types implementing `Blend`, you can start a new
+/// blend operation using [`Blendable::blend_additive`] or [`Blendable::blend_interp`].
+pub struct Blender<T>
+where
+    T: Blend,
+{
+    /// The value being blended.
+    value: T,
+    /// The cumulative weight of the blend operation so far. Every new blend operation
+    /// increases this weight.
+    weight: f32,
+}
+
+impl<T> Blender<T>
+where
+    T: Blend,
+{
+    /// Addatively blends the value into the blender.
+    pub fn blend_additive(self, value: T, weight: f32) -> Self {
+        Blender {
+            value: T::blend(self.value, value, weight),
+            weight: self.weight + weight,
+        }
+    }
+
+    /// Interpolatively blends the value into the blender.
+    pub fn blend_interp(self, value: T, weight: f32) -> Self {
+        let cumulative_weight = self.weight + weight;
+        Blender {
+            value: T::interp(&self.value, &value, weight / cumulative_weight),
+            weight: cumulative_weight,
+        }
+    }
+
+    /// Finishes the blend and returns the blended value.
+    pub fn finish(self) -> T {
+        self.value
+    }
+}
+
+/// This is an extension trait to `Blend` with methods to easily blend and interpolate
+/// between multiple values.
+pub trait Blendable: Blend {
+    /// Addatively blends the value into a new blender.
+    fn blend_additive(self, weight: f32) -> Blender<Self> {
+        Blender {
+            value: Self::blend(Self::IDENTITY, self, weight),
+            weight,
+        }
+    }
+
+    /// Interpolateively blends the value into a new blender.
+    fn blend_interp(self, weight: f32) -> Blender<Self> {
+        Blender {
+            value: self,
+            weight,
+        }
+    }
+}
+
+impl<T> Blendable for T where T: Blend {}
+
+macro_rules! impl_blendable_vectorspace {
+    ($ty: ident) => {
+        impl Blend for $ty {
+            const IDENTITY: Self = <$ty as VectorSpace>::ZERO;
+
+            #[inline]
+            fn blend(self, other: Self, blend_weight: f32) -> Self {
+                self + other * blend_weight
+            }
+        }
+    };
+}
+
+impl_blendable_vectorspace!(f32);
+impl_blendable_vectorspace!(Vec2);
+impl_blendable_vectorspace!(Vec3);
+impl_blendable_vectorspace!(Vec3A);
+impl_blendable_vectorspace!(Vec4);
+
+macro_rules! impl_blendable_color {
+    ($ty: ident) => {
+        impl Blend for $ty {
+            const IDENTITY: Self = $ty::BLACK;
+
+            #[inline]
+            fn blend(self, other: Self, blend_weight: f32) -> Self {
+                $ty::from_vec4(self.to_vec4() + other.to_vec4() * blend_weight)
```

Comment:
Isn't this supposed to be a convex (or at least an affine) combination?
```suggestion
                $ty::from_vec4(self.to_vec4() * (1.0 - blend_weight) + other.to_vec4() * blend_weight)
```
Also I wonder if we want to blend alpha like that.

Reply:
nooo `Blend` is for additive blending, which is just a linear combination. if we did it this way, `blend` would be equivalent to `interpolate`. take a closer look at `Blender<T>::blend_additive` and `Blender<T>::blend_interp`.

Reply:
this is a pretty direct translation of the math as it was done previously (just with the extra complexity of `Animatable` cut out).

Reply:
but it's an interesting idea to alpha-blend here. additive blending is all about layering effects on top of each other, and alpha blending does sort of capture that for colors. @mweatherley @viridia should additive blending for animated colors involve alpha blending?

Reply:
I don't think there's an obvious right answer here. If I add A to B, does the alpha in A affect the blend factor, or does it affect the result alpha? My first question would be, "what does photoshop / gimp / inkscape / CSS do?"

I would check out what linebender does here in their new `color` crate (which was partly based on bevy_color) - they are far more pedantic/strict about color spaces than we are: https://docs.rs/color/0.2.3/src/color/dynamic.rs.html#412-430

Just FYI, it's interesting that linebender's `color` crate defines premultiplied-alpha colors as distinct types (using a generic parameter). This was an idea that we discussed early on, and initially I was against it (I thought it would make the color classes too complicated) but now I am not so sure that was the right choice.

Reply:
it seems as though the [current implementation](https://github.com/bevyengine/bevy/blob/5973ba418f9fd7aa7e255b59c48b8ab2bf64850a/crates/bevy_animation/src/animatable.rs#L57) just does a linear combination including alpha, without considering premultiplication or alpha blending. This PR at least consistent with the current behavior.

Let's mark color blend-mode support and possible typed alpha state as problems for a future PR.